Shell/Builtin.cpp

```diff
@@ -0,0 +1,743 @@
+        return 1;
+
+    // -v implies -p
+    print = print || number_when_printing;
+
+    if (clear) {
+        for (size_t i = 1; i < directory_stack.size(); i++)
+            directory_stack.remove(i);
+    }
+
+    for (auto& path : paths)
+        directory_stack.append(path);
+
+    if (print) {
+        auto idx = 0;
+        for (auto& directory : directory_stack) {
+            if (number_when_printing)
+                printf("%d ", idx++);
+            print_path(directory);
+            fputc('\n', stdout);
+        }
+    }
+
+    return 0;
+}
+
+int Shell::builtin_exit(int, const char**)
+{
+    if (!jobs.is_empty()) {
+        if (!m_should_ignore_jobs_on_next_exit) {
+            printf("Shell: Hey dude, you have %zu active job%s, run 'exit' again to really exit.\n", jobs.size(), jobs.size() > 1 ? "s" : "");
+            m_should_ignore_jobs_on_next_exit = true;
+            return 1;
+        }
+    }
+    stop_all_jobs();
+    save_history();
+    printf("Good-bye!\n");
+    exit(0);
+    return 0;
+}
+
+int Shell::builtin_export(int argc, const char** argv)
+{
+    Vector<const char*> vars;
+
+    Core::ArgsParser parser;
+    parser.add_positional_argument(vars, "List of variable[=value]'s", "values", Core::ArgsParser::Required::No);
+
+    if (!parser.parse(argc, const_cast<char**>(argv), false))
+        return 1;
+
+    if (vars.size() == 0) {
+        for (int i = 0; environ[i]; ++i)
+            puts(environ[i]);
+        return 0;
+    }
+
+    int return_value = 0;
+
+    for (auto& value : vars) {
+        auto parts = String { value }.split_limit('=', 2);
+
+        if (parts.size() == 1) {
+            auto values = lookup_local_variable(parts[0])->resolve_as_list(*this);
+            StringBuilder builder;
+            builder.join(" ", values);
+            parts.append(builder.to_string());
+        }
+
+        int setenv_return = setenv(parts[0].characters(), parts[1].characters(), 1);
+
+        if (setenv_return != 0) {
+            perror("setenv");
+            return_value = 1;
+            break;
+        }
+
+        if (parts[0] == "PATH")
+            cache_path();
+    }
+
+    return return_value;
+}
+
+int Shell::builtin_fg(int argc, const char** argv)
+{
+    int job_id = -1;
+
+    Core::ArgsParser parser;
+    parser.add_positional_argument(job_id, "Job id to bring to foreground", "job_id", Core::ArgsParser::Required::No);
+
+    if (!parser.parse(argc, const_cast<char**>(argv), false))
+        return 1;
+
+    if (job_id == -1 && !jobs.is_empty())
+        job_id = find_last_job_id();
+
+    Job* job = nullptr;
+
+    for (auto& entry : jobs) {
+        if (entry.value->job_id() == (u64)job_id) {
+            job = entry.value;
+            break;
+        }
+    }
+    if (!job) {
+        if (job_id == -1) {
+            printf("fg: no current job\n");
+        } else {
+            printf("fg: job with id %d not found\n", job_id);
+        }
+        return 1;
+    }
```

Comment:
Perhaps all this job lookup logic could be shared?

Reply:
Oddly enough, it's already there, it just needs to be used here.